source/agora/script/Engine.d

```diff
@@ -735,6 +755,144 @@ public class Engine
+                ~ " opcode requires 4 items on the stack";
+            return err1;
+        }
+
+        const min_seq_bytes = stack.pop();
+        if (min_seq_bytes.length != ulong.sizeof)
+        {
+            static immutable err2 = op.to!string
+                ~ " opcode requires 8-byte minimum sequence on the stack";
+            return err2;
+        }
+
+        const ulong min_sequence = littleEndianToNative!ulong(
+            min_seq_bytes[0 .. ulong.sizeof]);
+
+        const key_bytes = stack.pop();
+        if (key_bytes.length != Point.sizeof)
+        {
+            static immutable err3 = op.to!string
+                ~ " opcode requires 32-byte public key on the stack";
+            return err3;
+        }
+
+        const Point point = Point(key_bytes);
+        if (!point.isValid())
+        {
+            static immutable err4 = op.to!string
+                ~ " 32-byte public key on the stack is invalid";
+            return err4;
+        }
+
+        const seq_bytes = stack.pop();
+        if (seq_bytes.length != ulong.sizeof)
+        {
+            static immutable err5 = op.to!string
+                ~ " opcode requires 8-byte sequence on the stack";
+            return err5;
+        }
+
+        const ulong sequence = littleEndianToNative!ulong(
+            seq_bytes[0 .. ulong.sizeof]);
+        if (sequence < min_sequence)
+        {
+            static immutable err6 = op.to!string
+                ~ " sequence is not equal to or greater than min_sequence";
+            return err6;
+        }
+
+        const sig_bytes = stack.pop();
+        if (sig_bytes.length != Signature.sizeof)
+        {
+            static immutable err7 = op.to!string
+                ~ " opcode requires 64-byte signature on the stack";
+            return err7;
+        }
+
+        const sig = Signature(sig_bytes);
+
+        // workaround: input index not explicitly passed in
+        import std.algorithm : countUntil;
+        const long input_idx = tx.inputs.countUntil(input);
+        assert(input_idx != -1, "Input does not belong to this transaction");
+
+        const Hash challenge = getSequenceChallenge(tx, sequence, input_idx);
+        sig_valid = Schnorr.verify(point, sig, challenge);
+        return null;
+    }
+}
+
+/*******************************************************************************
+
+    Gets the challenge hash for the provided transaction, sequence ID.
+
+    Params:
+        tx = the transaction to sign
+        sequence = the sequence ID to hash
+        input_idx = the associated input index we're signing for
+
+    Returns:
+        the challenge as a hash
+
+*******************************************************************************/
+
+public Hash getSequenceChallenge (in Transaction tx, in ulong sequence,
+    in ulong input_idx) nothrow @safe
+{
+    assert(input_idx < tx.inputs.length, "Input index is out of range");
+
+    Transaction dup;
```

Comment:
The variable name `dup` is confusing with the `dup` function.

Reply:
Good point. I'll rename it in another PR.